### ISSUE TITLE
Define plugin composition

### ISSUE THREAD

**Issue by phuu** [phuu/fetch-engine#3]
Issue #1 threw up a question about combining plugins, or only applying them to a subset of requests. It's desirable that an existing composition can be further composed.

Each plugin needs to be able to 'drop' the request, and prevent it from going any further. This lends itself well to the 'then' pattern that JSers have learnt from Promises. Example to follow...

**Comment by phuu** [comment]
An example, in slightly verbose JS...

```js
class Some {
  constructor(v) {
    this.type = Some.type;
    this.value = v;
  }
}
Some.type = 'Some';

class None {
  constructor(v) {
    this.type = None.type;
  }
}
None.type = 'None';

const Maybe = { Some, None };

const composeP = (f, g) => {
  return x => {
    const y = g(x);
    return (y.type === Maybe.Some.type
            ? f(y.value)
            : y);
  };
};

const filterRequest = r =>
  (r.path.startsWith('/boop')
   ? new Maybe.Some(r)
   : new Maybe.None());


const isInFlight = r => false; // TODO!
const dedupeRequest = r => {
  if (isInFlight(r)) {
    return new None();
  }
  return new Some(r);
};

const boopDedupe = composeP(dedupeRequest, filterRequest);
```

Head [over here](https://jsbin.com/bideza/3/edit?js,console) to see it in action.

**Comment by phuu** [comment]
While this is conceptually compatible with Promises, it might as well just *be* Promises. Will work on a sketch.

**Comment by phuu** [comment]
Closing, as this is now better addressed in #1.
